### ISSUE TITLE
warning C4834 since VS2017 15.6 update 

### ISSUE THREAD

**Issue by luckyluc74** [chriskohlhoff/asio#290]
Before the update from visual studio 2017 15.5 to 15.6 I had no warnings but after the update I get the following warnings. Should I ignore this warning or must there be something changed in asio? FYI i am using c++17.  

```
\asio-asio-1-10-8\asio\include\asio\buffer.hpp(530): warning C4834: discarding return value of function with 'nodiscard' attribute
\asio-asio-1-10-8\asio\include\asio\buffer.hpp(529): note: while compiling class template member function 'void asio::detail::buffer_debug_check<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>::operator ()(void)'
2>        with
2>        [
2>            _Ty=asio::detail::buffered_stream_storage::byte_type
2>        ]
2>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.13.26128\include\type_traits(16707566): note: see reference to function template instantiation 'void asio::detail::buffer_debug_check<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>::operator ()(void)' being compiled
2>        with
2>        [
2>            _Ty=asio::detail::buffered_stream_storage::byte_type
2>        ]
\asio-asio-1-10-8\asio\include\asio\buffer.hpp(1118): note: see reference to class template instantiation 'asio::detail::buffer_debug_check<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>' being compiled
2>        with
2>        [
2>            _Ty=asio::detail::buffered_stream_storage::byte_type
2>        ]
\asio-asio-1-10-8\asio\include\asio\detail\buffered_stream_storage.hpp(57): note: see reference to function template instantiation 'asio::mutable_buffers_1 asio::buffer<asio::detail::buffered_stream_storage::byte_type,std::allocator<_Ty>>(std::vector<_Ty,std::allocator<_Ty>> &)' being compiled
2>        with
2>        [
2>            _Ty=asio::detail::buffered_stream_storage::byte_type
2>        ]
```

**Comment by chronoxor** [comment]
I also found two more warnings generating in Visual Studio 2017 build 15.6 with /std:c++17 and /W4 /WX flags:

1) Deprecated std::result_of 
z:\projects\cppprojects\cppserver\modules\asio\asio\include\asio\detail\type_traits.hpp(56): warning C4996: 'std::result_of': warning STL4014: std::result_of and std::result_of_t are deprecated in C++17. They are superseded by std::invoke_result and std::
invoke_result_t. You can define _SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [Z:\Projects\CppProjects\CppServer\temp\modules\asio.vcxproj]
  c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.13.26128\include\type_traits(1809): note: see declaration of 'std::result_of'

2) Deprecated std::allocator<void>
z:\projects\cppprojects\cppserver\modules\asio\asio\include\asio\associated_allocator.hpp(102): warning C4996: 'std::allocator<void>': warning STL4009: std::allocator<void> is deprecated in C++17. You can define _SILENCE_CXX17_ALLOCATOR_VOID_DEPRECATION_W
ARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [Z:\Projects\CppProjects\CppServer\temp\modules\asio.vcxproj]
  c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.13.26128\include\xmemory0(1067): note: see declaration of 'std::allocator<void>'

**Comment by chriskohlhoff** [comment]
Should be fixed in bf70ad6062f85d23255f5e07c7d77cd36df12193 and b4245aac247881818af03d00c0f9b1e6f009cd31.

Won't fix `std::allocator<void>` warning because the warning is wrong. The `allocator<void>` template specialisation is deprecated but the `allocator<void>` template instantiation is not. Please disable the warning as suggested.
